beacon-chain/sync/regular_sync.go

```diff
@@ -0,0 +1,321 @@
+		beaconState.LastStateRecalculationSlot(),
+		block.SlotNumber(),
+	)
+	if err != nil {
+		log.Errorf("Failed to get proposer shard ID: %v", err)
+		return
+	}
+
+	// TODO(#258): stubbing public key with empty 32 bytes.
+	if err := attestation.VerifyProposerAttestation([32]byte{}, proposerShardID); err != nil {
+		log.Errorf("Failed to verify proposer attestation: %v", err)
+		return
+	}
+
+	_, sendAttestationSpan := trace.StartSpan(ctx, "sendAttestation")
+	log.WithField("attestationHash", fmt.Sprintf("%#x", attestation.Key())).Debug("Sending newly received attestation to subscribers")
+	rs.attestationService.IncomingAttestationFeed().Send(attestation)
+	sendAttestationSpan.End()
+
+	_, sendBlockSpan := trace.StartSpan(ctx, "sendBlock")
+	log.WithField("blockHash", fmt.Sprintf("%#x", blockHash)).Debug("Sending newly received block to subscribers")
+	rs.chainService.IncomingBlockFeed().Send(block)
+	sendBlockSpan.End()
+}
+
+// handleBlockRequestBySlot processes a block request from the p2p layer.
+// if found, the block is sent to the requesting peer.
+func (rs *RegularSync) handleBlockRequestBySlot(msg p2p.Message) {
+	ctx, blockRequestSpan := trace.StartSpan(msg.Ctx, "RegularSync_blockRequestBySlot")
+	defer blockRequestSpan.End()
+
+	request, ok := msg.Data.(*pb.BeaconBlockRequestBySlotNumber)
+	if !ok {
+		log.Error("Received malformed beacon block request p2p message")
+		return
+	}
+
+	ctx, getBlockSpan := trace.StartSpan(ctx, "getBlockBySlot")
+	block, err := rs.db.GetBlockBySlot(request.GetSlotNumber())
+	getBlockSpan.End()
+	if err != nil || block == nil {
+		log.Errorf("Error retrieving block from db: %v", err)
+		return
+	}
+
+	_, sendBlockSpan := trace.StartSpan(ctx, "sendBlock")
+	log.WithField("slotNumber", fmt.Sprintf("%d", request.GetSlotNumber())).Debug("Sending requested block to peer")
+	rs.p2p.Send(block.Proto(), msg.Peer)
+	sendBlockSpan.End()
+}
+
+func (rs *RegularSync) handleChainHeadRequest(msg p2p.Message) {
+	if _, ok := msg.Data.(*pb.ChainHeadRequest); !ok {
+		log.Errorf("message is of the incorrect type")
+		return
+	}
+
+	block, err := rs.db.GetChainHead()
+	if err != nil {
+		log.Errorf("Could not retrieve chain head %v", err)
+		return
+	}
+
+	hash, err := block.Hash()
+	if err != nil {
+		log.Errorf("Could not hash block %v", err)
+		return
+	}
+
+	req := &pb.ChainHeadResponse{
+		Slot:  block.SlotNumber(),
+		Hash:  hash[:],
+		Block: block.Proto(),
+	}
+
+	rs.p2p.Send(req, msg.Peer)
+}
+
+// receiveAttestation accepts an broadcasted attestation from the p2p layer,
+// discard the attestation if we have gotten before, send it to attestation
+// service if we have not.
+func (rs *RegularSync) receiveAttestation(msg p2p.Message) {
+	data := msg.Data.(*pb.AggregatedAttestation)
+	a := types.NewAttestation(data)
+	h := a.Key()
+
+	attestation, err := rs.db.GetAttestation(h)
+	if err != nil {
+		log.Errorf("Could not check for attestation in DB: %v", err)
+		return
+	}
+	if attestation != nil {
```

Comment:
So if it is nil and validator does not exist, we'll be forwarding empty data via the feed? should it return instead?

Reply:
I think we forward the data to the feed, if the attestation does not have the validator, then we have to process the attestation, no ?